Upstream: https://gitlab.gnome.org/GNOME/at-spi2-core/-/commit/d442ee182ec8fa095c6bc5298a17663cfc70cf9a
From: Mike Gorse <mgorse@suse.com>
Date: Fri, 17 Jan 2025 13:57:51 -0600
Subject: [PATCH] atk-adaptor: Have the SpiCache keep its own weak ref

This makes the code less convoluted. Previously, SpiCache relied on the
code in accessible-register.c to send a signal when the object went away. At
least in theory, this could cause a dangling pointer if the root
accessible changes or goes away, since it has a fixed DBus object path and
is thus not registered in SpiRegister. We have a couple of crashes involving
AT-SPI that are hard to reproduce or debug (#176, #178), and it seems
better to simplify the code.

The SpiRegister now only emits object-deregistered when an object is
marked defunct, not when it is deregistered via the weak ref. SpiCache is
its only consumer and is now notified directly when an AtkObject is being
finalized.
---
 atk-adaptor/accessible-cache.c    | 21 +++++++++++++++++----
 atk-adaptor/accessible-register.c | 12 +++++++-----
 2 files changed, 24 insertions(+), 9 deletions(-)

diff --git a/atk-adaptor/accessible-cache.c b/atk-adaptor/accessible-cache.c
index 1452a109..26d9f75e 100644
--- a/atk-adaptor/accessible-cache.c
+++ b/atk-adaptor/accessible-cache.c
@@ -133,6 +133,20 @@ spi_cache_init (SpiCache *cache)
                     (GCallback) toplevel_added_listener, NULL);
 }
 
+static void
+cache_weak_ref (gpointer data, GObject *gobj)
+{
+  SpiCache *cache = SPI_CACHE (data);
+
+  g_hash_table_remove (cache->objects, gobj);
+}
+
+static void
+spi_cache_remove_weak_ref (gpointer key, gpointer val, gpointer cache)
+{
+  g_object_weak_unref (val, cache_weak_ref, cache);
+}
+
 static void
 spi_cache_finalize (GObject *object)
 {
@@ -141,6 +155,7 @@ spi_cache_finalize (GObject *object)
   while (!g_queue_is_empty (cache->add_traversal))
     g_object_unref (G_OBJECT (g_queue_pop_head (cache->add_traversal)));
   g_queue_free (cache->add_traversal);
+  g_hash_table_foreach (cache->objects, spi_cache_remove_weak_ref, cache);
   g_hash_table_unref (cache->objects);
 
   g_signal_handlers_disconnect_by_func (spi_global_register,
@@ -169,6 +184,7 @@ remove_object (GObject *source, GObject *gobj, gpointer data)
                spi_register_object_to_path (spi_global_register, gobj));
 #endif
       g_signal_emit (cache, cache_signals[OBJECT_REMOVED], 0, gobj);
+      g_object_weak_unref (G_OBJECT (gobj), cache_weak_ref, cache);
       g_hash_table_remove (cache->objects, gobj);
     }
   else if (g_queue_remove (cache->add_traversal, gobj))
@@ -183,6 +199,7 @@ add_object (SpiCache *cache, GObject *gobj)
   g_return_if_fail (G_IS_OBJECT (gobj));
 
   g_hash_table_insert (cache->objects, gobj, NULL);
+  g_object_weak_ref (G_OBJECT (gobj), cache_weak_ref, cache);
 
 #ifdef SPI_ATK_DEBUG
   g_debug ("CACHE ADD - %s - %d - %s\n", atk_object_get_name (ATK_OBJECT (gobj)),
@@ -303,10 +320,6 @@ add_pending_items (gpointer data)
     {
       current = g_queue_pop_head (to_add);
 
-      /* Make sure object is registerd so we are notified if it goes away */
-      g_free (spi_register_object_to_path (spi_global_register,
-                                           G_OBJECT (current)));
-
       add_object (cache, G_OBJECT (current));
       g_object_unref (G_OBJECT (current));
     }
diff --git a/atk-adaptor/accessible-register.c b/atk-adaptor/accessible-register.c
index 6361ec46..e7521c0c 100644
--- a/atk-adaptor/accessible-register.c
+++ b/atk-adaptor/accessible-register.c
@@ -198,12 +198,14 @@ spi_register_deregister_object (SpiRegister *reg, GObject *gobj, gboolean unref)
   ref = object_to_ref (gobj);
   if (ref != 0)
     {
-      g_signal_emit (reg,
-                     register_signals[OBJECT_DEREGISTERED],
-                     0,
-                     gobj);
       if (unref)
-        g_object_weak_unref (gobj, deregister_object, reg);
+        {
+          g_signal_emit (reg,
+                         register_signals[OBJECT_DEREGISTERED],
+                         0,
+                         gobj);
+          g_object_weak_unref (gobj, deregister_object, reg);
+        }
       g_hash_table_remove (reg->ref2ptr, GINT_TO_POINTER (ref));
 
 #ifdef SPI_ATK_DEBUG
